### DISCUSSION TITLE
Custom Registry for NPM

### DISCUSSION THREAD

**Discussion by juancarlosjr97** [renovatebot/renovate#22916] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.133.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Hello,

We have internally where I work JFrog as our registry, we started rolling Renovate to other projects consuming internally modules to our JFrog registry

### What you tried so far.

 We have added the following hostRules to our config.js

```
const npmHostRules = {
  hostType: "npm",
  matchHost: "https://*****.jfrog.io/*****/api/npm/npm/",
  username: ****,
  password: ****,
};

const disableNpmJs = {
  matchHost: "registry.npmjs.org",
  enabled: false,
};
```

Unfortunately, I cannot share the log, sensitive information, and the matchHost is correct as that is what we used in our npmrc.

Renovate fails as I disabled npmjs, for those packages that are in the public registry of npmjs and in the private work, but those only internally failed, but after disabling just fail entirely.

Reading other issues, I found out about npmrc that can be added to Renovate configuration https://docs.renovatebot.com/getting-started/private-packages/#add-npmrc-string-to-renovate-config and all our package-lock.json are pointing to our JFrog registry.

Does that mean I need to add the npmrc to Renovate to work for this to work or should work with the about host rule and something else could be wrong?

Thank you.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
Host rules are insufficient on their own. You need npmrc or packageRules, please read the docs on these config options

**Reply by juancarlosjr97** [reply]
Thank you so much!

I have added the following configuration

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "group:all"
  ],
  "enabledManagers": [
    "docker-compose",
    "dockerfile",
    "nodenv",
    "npm",
    "nvm",
    "pre-commit",
    "regex"
  ],
  "hostRules": [
    {
      "matchHost": "registry.npmjs.org",
      "enabled": "false"
    },
    {
      "matchHost": "private.jfrog.io",
      "hostType": "npm"
    }
  ],
  "dependencyDashboard": true,
  "npmrc": "registry=https://private.jfrog.io/artifactory/api/npm/npm/\n@private:registry=https://private.jfrog.io/artifactory/api/npm/npm/\n@private:registry=https://private.jfrog.io/artifactory/api/npm/npm/\n//private.jfrog.io/artifactory/api/npm/npm/:_password=***\n//private.jfrog.io/artifactory/api/npm/npm/:username=juancarlosblancodelgado\n//private.jfrog.io/artifactory/api/npm/npm/:email=juancarlos.blancodelgado@private.com\n//private.jfrog.io/artifactory/api/npm/npm/:always-auth=true"
}

```

It fails with the following error 401 from the private registry, when runs on GitLab self-hosted that uses a docker-in-docker, but when I test the same from GitHub it works.

The DEBUG shows this error

`"https://private.jfrog.io/private/api/npm/npm/ansi-regex (GET,401)": 1,`

I wonder if it is the docker-in-docker causing the failure missing the npmrc.

I have attached the full logs as I cannot see to see the error as just fails the request to the private registry. Other jobs using the same configuration work publishing or pulling modules from that private registry.

I have tried different ways the npmrc, registry first, then the auth, with password, username, etc, following JFrog setup https://jfrog.com/help/r/jfrog-artifactory-documentation/npm-registry for the npmrc, and I also used the exactly one I used locally as it is the same node and npm version.

Important note, the debugging log I am attaching has the full configuration I am using, but using the simple one above or full configuration we used to manage renovate, it fails with the same error, in addition to removing any business related, so paths, project names, directories, and any other thing that could identify internally our system has been updated with dummy information but the important information is there.


### Relevant debug logs

<details><summary>Logs</summary>

```

```

</details>

**Reply by juancarlosjr97** [reply]
So, I have removed the logs as I have a solution to the issue but not a reason why, so I am testing it little by little. It was a configuration or a npmrc problem, unclear yet but it is working as it should pulling the dependencies from the private registry. As soon as I have it working again, I will post the reason as I am enabling configuration by configuration to identify which one was causing the issue.

Thank you!

**Reply by viceice** [reply]
for npm the hostrule url must match the registry url from npmrc exactly and it should always include a training slash
